northstar/src/runtime/pipe.rs

```diff
@@ -0,0 +1,256 @@
+// Copyright (c) 2019 - 2020 ESRLabs
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+use nix::{fcntl, fcntl::OFlag, unistd};
+use std::{
+    convert::TryFrom,
+    io,
+    os::unix::io::{AsRawFd, RawFd},
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::io::{unix::AsyncFd, AsyncRead, AsyncWrite, ReadBuf};
+
+/// Opens a pipe(2) with both ends blocking
+pub(crate) fn pipe() -> io::Result<(PipeReader, PipeWriter)> {
+    let (readfd, writefd) = unistd::pipe().map_err(from_nix)?;
+    Ok((PipeReader(readfd), PipeWriter(writefd)))
+}
+
+/// Read end of a pipe(2)
+#[derive(Debug)]
+pub(crate) struct PipeReader(RawFd);
+
+impl std::io::Read for PipeReader {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        unistd::read(self.0, buf).map_err(from_nix)
+    }
+}
+
+impl AsRawFd for PipeReader {
+    fn as_raw_fd(&self) -> RawFd {
+        self.0
+    }
+}
+
+impl Drop for PipeReader {
+    fn drop(&mut self) {
+        if let Err(e) = unistd::close(self.0) {
+            log::error!("Failed to close pipe: {}", e);
```

Comment:
There's not much we can do when the close fails. I'd panic instead of just logging it. Retrying doesn't make sense.